src/wp-includes/class-wp-scripts.php

```diff
@@ -295,14 +304,34 @@ public function do_item( $handle, $group = false ) {
 		}
 
 		$before_handle = $this->print_inline_script( $handle, 'before', false );
-		$after_handle  = $this->print_inline_script( $handle, 'after', false );
 
 		if ( $before_handle ) {
 			$before_handle = sprintf( "<script%s id='%s-js-before'>\n%s\n</script>\n", $this->type_attr, esc_attr( $handle ), $before_handle );
 		}
 
+		// Eligible loading strategies will only be 'async', 'defer', or ''.
+		$strategy     = $this->get_eligible_loading_strategy( $handle );
+		$after        = ( '' === $strategy ) ? 'after' : 'after-standalone';
+		$after_handle = $this->print_inline_script( $handle, $after, false );
+
 		if ( $after_handle ) {
-			$after_handle = sprintf( "<script%s id='%s-js-after'>\n%s\n</script>\n", $this->type_attr, esc_attr( $handle ), $after_handle );
+			$after_handle = sprintf(
+				"<script%1\$s id='%2\$s-js-after'>\n%3\$s\n</script>\n",
+				$this->type_attr,
+				esc_attr( $handle ),
+				$after_handle
+			);
+		}
+		if ( '' !== $strategy ) {
+			$after_non_standalone_handle = $this->print_inline_script( $handle, 'after-non-standalone', false );
+
+			if ( $after_non_standalone_handle ) {
+				$after_handle .= sprintf(
+					"<script type='text/template' id='%1\$s-js-after' data-wp-executes-after='%1\$s'>\n%2\$s\n</script>\n",
+					esc_attr( $handle ),
+					$after_non_standalone_handle
+				);
+			}
```

Comment:
See my comment on  the `add_inline_script()` method: I think this approach is error-prone and overly complicated since it mixes the position and whether inline scripts are standalone.

I think with what I'm proposing in that related comment, we could keep the `print_inline_script()` calls in here exactly as before, and only when wrapping them in a `script` tag we would need to check whether or not this group of inline scripts should be printed as standalone or not.